## Laboratorio 3 - Clasificación

## Paula Hípola Gómez, José Ignacio Navas Sanz y Belén Ortega Pérez

## Entregable 1 - Clasificador de Flores

In [1]:
%config IPCompleter.greedy=True
%reset
#Importamos las librerías necesarias
#Numpy
import numpy as np 
#Pandas
import pandas as pd
#Scikit-Learn
import sklearn as sk
#Matplotlib
import matplotlib.pyplot as plt

Once deleted, variables cannot be recovered. Proceed (y/[n])? y


## Carga y Visualización de Datos

In [2]:
#Importamos el CSV
dataset = pd.read_csv('iris.csv')
#Mostramos los datos
dataset

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris-virginica  
149  Iris-virginica  

[150 rows x 6 columns]

In [89]:
dataset.describe()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count  150.000000     150.000000    150.000000     150.000000    150.000000
mean    75.500000       5.843333      3.054000       3.758667      1.198667
std     43.445368       0.828066      0.433594       1.764420      0.763161
min      1.000000       4.300000      2.000000       1.000000      0.100000
25%     38.250000       5.100000      2.800000       1.600000      0.300000
50%     75.500000       5.800000      3.000000       4.350000      1.300000
75%    112.750000       6.400000      3.300000       5.100000      1.800000
max    150.000000       7.900000      4.400000       6.900000      2.500000

### 1) Haz todo el preprocesamiento para crear un set de entrenamiento y otro de validación que permitan crear un clasificador atendiendo sólo a las características necesarias. Explica qué has hecho y porqué. (0,5 puntos)

Necesitamos separar unas instancias de entrenamiento y otras instancias para validación. La división típica suele ser un 80-20 si bien podemos intentar apurar hasta un 75-25 o incluso 70-30 siendo el número más grande el que representa el procentaje del conjunto de entrenamiento. El problema de definir muchos datos para entrenar y pocos para validar se denomina como "overfitting" y nos puede producir un sobreajuste de las instancias del set de validación. Esto no siempre es interesante ya que no estaríamos obteniendo un modelo realista. El ajuste contrario se denomina "underfitting", y del mismo modo que el "overfitting", no produce un modelo fiable al ser menos preciso que la realidad. Necesitamos además separar la clase objetivo, que en este caso será la "Especie"

Lo primero que hemos tenido que hacer, es aleatorizar el dataset ya que los datos vienen ordenados por la clase objetivo, que en nuestro caso es la especie de las flores. De este modo, si no lo reordenásemos de modo aleatorio, al realizar la división 80-20 que hemos pensado, estaríamos entrenando con dos clases objetivo, mientras que para la validación estaríamos buscando solo en una clase, por lo que el set de entrenamiento no quedará bien entrenado pudiendo producir errores de identificación.

In [3]:
X = dataset.iloc[:, 1:5].values
y = dataset.iloc[:, 5].values

In [4]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [6]:
pd.DataFrame(X_train)

0    1    2    3
0    6.4  3.1  5.5  1.8
1    5.4  3.0  4.5  1.5
2    5.2  3.5  1.5  0.2
3    6.1  3.0  4.9  1.8
4    6.4  2.8  5.6  2.2
..   ...  ...  ...  ...
115  4.9  3.1  1.5  0.1
116  6.3  2.9  5.6  1.8
117  5.8  2.7  4.1  1.0
118  7.7  3.8  6.7  2.2
119  4.6  3.2  1.4  0.2

[120 rows x 4 columns]

In [7]:
pd.DataFrame(X_test)

0    1    2    3
0   5.8  2.8  5.1  2.4
1   6.0  2.2  4.0  1.0
2   5.5  4.2  1.4  0.2
3   7.3  2.9  6.3  1.8
4   5.0  3.4  1.5  0.2
5   6.3  3.3  6.0  2.5
6   5.0  3.5  1.3  0.3
7   6.7  3.1  4.7  1.5
8   6.8  2.8  4.8  1.4
9   6.1  2.8  4.0  1.3
10  6.1  2.6  5.6  1.4
11  6.4  3.2  4.5  1.5
12  6.1  2.8  4.7  1.2
13  6.5  2.8  4.6  1.5
14  6.1  2.9  4.7  1.4
15  4.9  3.1  1.5  0.1
16  6.0  2.9  4.5  1.5
17  5.5  2.6  4.4  1.2
18  4.8  3.0  1.4  0.3
19  5.4  3.9  1.3  0.4
20  5.6  2.8  4.9  2.0
21  5.6  3.0  4.5  1.5
22  4.8  3.4  1.9  0.2
23  4.4  2.9  1.4  0.2
24  6.2  2.8  4.8  1.8
25  4.6  3.6  1.0  0.2
26  5.1  3.8  1.9  0.4
27  6.2  2.9  4.3  1.3
28  5.0  2.3  3.3  1.0
29  5.0  3.4  1.6  0.4

In [8]:
pd.DataFrame(y_train)

0
0    2
1    1
2    0
3    2
4    2
..  ..
115  0
116  2
117  1
118  2
119  0

[120 rows x 1 columns]

In [9]:
pd.DataFrame(y_test)

0
0   2
1   1
2   0
3   2
4   0
5   2
6   0
7   1
8   1
9   1
10  2
11  1
12  1
13  1
14  1
15  0
16  1
17  1
18  0
19  0
20  2
21  1
22  0
23  0
24  2
25  0
26  0
27  1
28  1
29  0

### 2) Prueba con distintas configuraciones de las dos métricas principales. La primera métrica corresponde al número de individuos que usarás para clasificar una nueva instancia y la segunda cómo vas a medir la cercanía de esa nueva instancia con el resto. ¿Qué decisiones has tomado? ¿Por qué? (1 punto)

La primera métrica que tenemos que configurar es el número de vecinos a decidir. En nuestro caso, vamos a optar por coger 5 vecinos puesto que la longitud de nuestro set de validación es 30, y la raíz cuadrada de 30 es 5.47. ¿Qué conseguimos con esta división? Que el número de vecinos sea lo más próximo a todos y cada uno de los registros de la manera más equidistante posible.

In [36]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.model_selection import cross_val_score

matricesManh = []
precisionM = []
matricesEuc = []
precisionE = []

for x in range(1,10):
    print('K=',x)
    #Manhattan
    knnManh = KNeighborsClassifier(n_neighbors=x,metric='manhattan')
    knnManh.fit(X_train, y_train)
    predictManh = knnEuclidean.predict(X_test)
    matrizConfManh = confusion_matrix(y_test, predictManh)
    matricesManh.append(matrizConfManh)
    accuracyM = accuracy_score(y_test, predictManh)*100
    precisionM.append(accuracyM)
    #Euclidean
    knnEuclidean = KNeighborsClassifier(n_neighbors=x,metric='euclidean')
    knnEuclidean.fit(X_train, y_train)
    predictEuclidean = knnEuclidean.predict(X_test)
    matrizConfEuc = confusion_matrix(y_test, predictEuclidean)
    matricesEuc.append(matrizConfEuc)
    accuracyE = accuracy_score(y_test, predictEuclidean)*100
    precisionE.append(accuracyE)
    #Resultados
    print('La precisión del modelo con distancia Manhattan es del: ' + str(round(accuracyM, 2)) + ' %.')
    print('La precisión del modelo con distancia Euclidean es del: ' + str(round(accuracyE, 2)) + ' %.')
    print('---------------------------------------------------------')
    
print('Precision media de Manhattan:', round(np.mean(precisionM),2))
print('Precision media de Euclidean:', round(np.mean(precisionE),2))

K= 1
La precisión del modelo con distancia Manhattan es del: 100.0 %.
La precisión del modelo con distancia Euclidean es del: 100.0 %.
---------------------------------------------------------
K= 2
La precisión del modelo con distancia Manhattan es del: 100.0 %.
La precisión del modelo con distancia Euclidean es del: 96.67 %.
---------------------------------------------------------
K= 3
La precisión del modelo con distancia Manhattan es del: 96.67 %.
La precisión del modelo con distancia Euclidean es del: 96.67 %.
---------------------------------------------------------
K= 4
La precisión del modelo con distancia Manhattan es del: 96.67 %.
La precisión del modelo con distancia Euclidean es del: 100.0 %.
---------------------------------------------------------
K= 5
La precisión del modelo con distancia Manhattan es del: 100.0 %.
La precisión del modelo con distancia Euclidean es del: 96.67 %.
---------------------------------------------------------
K= 6
La precisión del modelo con di

No esperabamos que ambas métricas fueran exactamente igual de precisas en su media. En este caso debemos elegir cuál es la que queremos. Nos decantamos por Manhattan porque nos da más precisión para el punto en el que queremos estudiar nosotros, que es K=5.

### 3) Elige la mejor configuración entre las anteriores. Para ello crea una tabla para ver cómo evoluciona la clasificación. Dibuja los resultados que se obtienen con ambas configuraciones elegidas como las mejores. (1 punto)

In [85]:
import matplotlib.pyplot as plt

metricas = ["euclidean", "manhattan", "minkowski"]
precisiones = []
errores = []
dataSet = []


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
    
for metrica in metricas:
      
    for x in range(1,10):
        modelo = KNeighborsClassifier(n_neighbors=x,metric=metrica)
        modelo.fit(X_train, y_train)
        
        # Cálculo de la precisión y el error.
        precision = round(accuracy_score(modelo.predict(X_test), y_test), 3) *100 
        error = round(100 - precision, 3) 
        precisiones.append(precision)
        errores.append(error)
        
        # Vamos a crear el Dataframe para la division
        data = {'K': x, 'Division': division, 'Métrica': metrica, 'Precisión': round(precision,3), 'Error': round(error,3)}
        dataSet.append(data)
        
dataFrameL = pd.DataFrame(dataSet)

In [86]:
dataFrameL.loc[dataFrameL['Métrica'] == 'euclidean'].describe()

K      Division  Precisión  Error
count  9.000000  9.000000e+00       9.00   9.00
mean   5.000000  1.000000e-01      98.90   1.10
std    2.738613  1.471962e-17       1.65   1.65
min    1.000000  1.000000e-01      96.70   0.00
25%    3.000000  1.000000e-01      96.70   0.00
50%    5.000000  1.000000e-01     100.00   0.00
75%    7.000000  1.000000e-01     100.00   3.30
max    9.000000  1.000000e-01     100.00   3.30

In [87]:
dataFrameL.loc[dataFrameL['Métrica'] == 'manhattan'].describe()

K      Division  Precisión  Error
count  9.000000  9.000000e+00       9.00   9.00
mean   5.000000  1.000000e-01      98.90   1.10
std    2.738613  1.471962e-17       1.65   1.65
min    1.000000  1.000000e-01      96.70   0.00
25%    3.000000  1.000000e-01      96.70   0.00
50%    5.000000  1.000000e-01     100.00   0.00
75%    7.000000  1.000000e-01     100.00   3.30
max    9.000000  1.000000e-01     100.00   3.30

In [88]:
dataFrameL.loc[dataFrameL['Métrica'] == 'minkowski'].describe()

K      Division  Precisión  Error
count  9.000000  9.000000e+00       9.00   9.00
mean   5.000000  1.000000e-01      98.90   1.10
std    2.738613  1.471962e-17       1.65   1.65
min    1.000000  1.000000e-01      96.70   0.00
25%    3.000000  1.000000e-01      96.70   0.00
50%    5.000000  1.000000e-01     100.00   0.00
75%    7.000000  1.000000e-01     100.00   3.30
max    9.000000  1.000000e-01     100.00   3.30

Hemos obtenido que para las métricas, Manhattan, Euclidean y Minkowski (por defecto en la librería de SciKit Learn), los valores de precisión y error son similares por lo que no nos difiere escoger una u otra. Esto quizá se debe a que la cantidad de datos no es muy grande, son 150 valores, de los que hemos seleccionado 120 para entrenar el modelo y 30 para predecir por lo que la tasa de precisión es normal que sea alta. No hemos estudiado a fondo cómo utiliza la distancia la métrica Minkowski, por lo que no tiene sentido escogerla para crear nuestro modelo. Nos decidimos por utilizar la métrica Manhattan, ya que como vimos en el punto 2, donde definimos nuestro valor óptima de K a 5, Manhattan era un poco más precisa que Euclidean. Por lo tanto, nuestro modelo será con K=5 y la métrica Manhattan.

In [79]:
KOptima = 5
modeloKNN = KNeighborsClassifier(n_neighbors=KOptima,metric='manhattan')
modeloKNN.fit(X_train, y_train)

KNeighborsClassifier(metric='manhattan')

### 4) Utiliza el clasificador para saber qué ocurre con las flores de un dataset que se llame “iris2clasif.csv “que obtendremos del dataset proporcionado. (0,5 puntos)

In [ ]:
#Probando el clasificador
